Problem 1

In [1]:
# All the imports
import pennylane as qml
import numpy as np

In [2]:
# Create the devices
dev1 = qml.device("default.qubit", wires=1)
dev2 = qml.device("default.qubit", wires=1)

In [3]:
# Create the circuit function, which takes
# Two angles: theta1 (θ_1) and theta2 (θ_2)
# Two functions: func1 and func2, where they can be either RX or RY
def circuit(theta1, theta2, func1, func2):
    func1(theta1, wires=0)
    func2(theta2, wires=0)
    return qml.expval(qml.PauliX(wires=0)), qml.probs(wires=0), qml.expval(qml.PauliZ(wires=0))

In [4]:
# Define the QNodes
@qml.qnode(dev1)
def circuit1(theta1, theta2):
    return circuit(theta1, theta2, qml.RX, qml.RY)

@qml.qnode(dev2)
def circuit2(theta1, theta2):
    return circuit(theta1, theta2, qml.RY, qml.RX)

In [5]:
# Drawing function
def draw_circuit(circuit, theta1, theta2):
    return qml.draw(circuit)(theta1, theta2)

In [6]:
# Initializations
theta1 = np.pi / 3
theta2 = np.pi / 4

# Compute results for both circuits
result1 = circuit1(theta1, theta2)
result2 = circuit2(theta2, theta1)

# Extract the results
expval1_x, probs1, expval1_z = result1
expval2_x, probs2, expval2_z = result2

In [7]:
# Draw the circuits and print the results
drawer1 = draw_circuit(circuit1, theta1, theta2)
print("Circuit 1: ",drawer1)
print(f"Circuit 1 - Expectation Value of Pauli-X: {expval1_x:.4f}")
print(f"Circuit 1 - Expectation Value of Pauli-Z: {expval1_z:.4f}")
print(f"Circuit 1 - Probability |0>: {probs1[0]:.4f}, Probability |1>: {probs1[1]:.4f}")

drawer2 = draw_circuit(circuit2, theta2, theta1)
print("Circuit 2: ",drawer2)
print(f"Circuit 2 - Expectation Value of Pauli-X: {expval2_x:.4f}")
print(f"Circuit 2 - Expectation Value of Pauli-Z: {expval2_z:.4f}")
print(f"Circuit 2 - Probability |0>: {probs2[0]:.4f}, Probability |1>: {probs2[1]:.4f}")

Circuit 1:  0: ──RX(1.05)──RY(0.79)─┤  <X>  Probs  <Z>
Circuit 1 - Expectation Value of Pauli-X: 0.3536
Circuit 1 - Expectation Value of Pauli-Z: 0.3536
Circuit 1 - Probability |0>: 0.6768, Probability |1>: 0.3232
Circuit 2:  0: ──RY(0.79)──RX(1.05)─┤  <X>  Probs  <Z>
Circuit 2 - Expectation Value of Pauli-X: 0.7071
Circuit 2 - Expectation Value of Pauli-Z: 0.3536
Circuit 2 - Probability |0>: 0.6768, Probability |1>: 0.3232


To check the expected value:

Let the final state of the |ψ>

Then, to get the expectation value of the Pauli-X gate, we need to calculate:<ψ|X|ψ>

In [8]:
# Calculate the absolute difference
absolute_difference_x = np.abs(expval1_x - expval2_x)
print(f"Absolute Difference of Pauli-X: {absolute_difference_x:.4f}")

absolute_difference_z = np.abs(expval1_z - expval2_z)
print(f"Absolute Difference of Paulit-Z: {absolute_difference_z:.4f}")

Absolute Difference of Pauli-X: 0.3536
Absolute Difference of Paulit-Z: 0.0000


Question 1:
What differences do you observe between the outcomes of the two circuits when
measuring in the Pauli-X basis?

Answer:

First, we need to note that the expectation value of Pauli-X obsevable is determined by the state's alignment with the X-axis on the Bloch sphere, meaning Pauli-X measures the projection of the final state along the X-axis. Inituitively, someone would say that we should have the same expectation value in both circuits. For example, let's take a classical example, suppose I have a ball that I should move only in two routines, either up then right, or right then up. Classically, someone would say that immaterial of which routine we take the ball will end at the same place, which is correct, but in classical scenarios, not in quantum mechanics. Quantum rotations don't commute, meaning the order of the applied unitary operations matters in quantum mechanics.

Going back to the question, assuming for the sack of clarification that theta1 = π/3 and theta2 = π/4, we observed that the expectation value of Pauli-X observable for circuit-1 is: 0.3536 and that of circuit-2 is: 0.7071. This means that circuit-2 has a stronger alignment with the X-axis. 


Question 2:
How does the result change when measuring in the Pauli-Z basis?

Answer:

Both circuits produce the same ⟨𝑍⟩ values. This is because the gate combinations in both circuits generate states with the same symmetry along the Z-axis., i.e. absolute difference = 0. Circuit-1: 0.3536 and Circuit-2: 0.3536, there absoulue diffecence = 0, as expected.

Question 3:
Does the order of operations (RX and RY) affect the results in the same way for
both Pauli-X and Pauli-Z measurements?

Answer:

No. The order in which RX and RY are applied significantly affects ⟨X⟩, but didn't affect that of ⟨𝑍⟩. Since, for ⟨X⟩ it results in different alignments with the X-axis, however for ⟨𝑍⟩ it results in a symmetric state indifferent of the order.

Question 4:
What insights can you derive about the nature of these rotation gates based on your results?

Answer:

Someone would notice that the probability of the computation basis (i.e., |0> and |1>) is the same for the 4 circuits (2 for ⟨X⟩, 2 for ⟨𝑍⟩), because the probability of these gates depends on their projection on the Z-axis, and since there were no rotations along that axis (only on X-axis and Y-axis),the effect of these rotations do not directly affect the state's alignment with the Z-axis in a way that changes the probabilities in the computational basis. Therefore, their probabilities are expected to be the same, and that what we observed. Circuit 1-2-3-4: P(|0>) = 0.6768 and P(|1>) = 0.3232. 